packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -0,0 +1,179 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui';
+
+import 'package:characters/characters.dart' show CharacterRange;
+
+import 'text_layout_metrics.dart';
+
+/// An interface for retrieving the logical text boundary (left-closed-right-open)
+/// at a given location in a document.
+///
+/// The input [TextPosition] points to a position between 2 code units (which
+/// can be visually represented by the caret if the selection were to collapse
+/// to that position).
+abstract class TextBoundary {
+  /// A constant constructor to enable subclass override.
+  const TextBoundary();
+
+  /// Returns the leading text boundary at the given location.
+  ///
+  /// The return value must be less or equal to the input position.
+  TextPosition getLeadingTextBoundaryAt(TextPosition position);
+
+  /// Returns the trailing text boundary at the given location, exclusive.
+  ///
+  /// The return value must be greater or equal to the input position.
+  TextPosition getTrailingTextBoundaryAt(TextPosition position);
+
+  /// Gets the text boundary range that encloses the input position.
+  TextRange getTextBoundaryAt(TextPosition position) {
+    return TextRange(
+      start: getLeadingTextBoundaryAt(position).offset,
+      end: getTrailingTextBoundaryAt(position).offset,
+    );
+  }
+}
+
+/// A text boundary that uses characters as logical boundaries.
+///
+/// This class takes grapheme clusters into account and avoid creating
+/// boundaries that generate malformed utf-16 characters.
+class CharacterBoundary extends TextBoundary {
+  /// Creates a [CharacterBoundary] with the text.
+  const CharacterBoundary(this._text);
+
+  final String _text;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    if (position.offset <= 0) {
+      return const TextPosition(offset: 0);
+    }
+    if (position.offset > _text.length ||
+        (position.offset == _text.length && position.affinity == TextAffinity.downstream)) {
+      return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    final int endOffset;
+    final int startOffset;
+    switch (position.affinity) {
+      case TextAffinity.upstream:
+        startOffset = math.min(position.offset - 1, _text.length);
```

Comment:
If the text is `ABC` and assume the paragraph is LTR and there's no soft wrapping, then `TextPosition(1, downstream)` and `TextPosition(1, upstream)` point to exactly the same caret position visually (at least based on the logic in `TextPainter`), but this class will return a different character range for these 2 ('B' with the downstream one and 'A' with the upstream one). I think it could be confusing to the app users that the same caret location (at least that's what appears to be the case for them) actually points to different characters.

Reply:
The "right" behavior to me here, if we want to account for text affinity, is probably to return the same range when `TextPosition(1, downstream)` and `TextPosition(1, upstream)` both give the same caret position. We'll have to query the text layout in order to do that.

Reply:
Discussed offline, `TextPosition(offset: x, affinity: upstream)` should be interpreted as pointing to the trailing half of the `x-1`th code unit, and `TextPosition(offset: x, affinity: downstream)` is the leading half of the `x`th code unit. Since this applies to all `TextBoundary` subclasses maybe we should explain that in `TextBoundary`'s documentation?

Reply:
done